source/extensions/http/stateful_session/header/header.cc

```diff
@@ -1,29 +1,79 @@
 #include "source/extensions/http/stateful_session/header/header.h"
 
+#include "absl/container/inlined_vector.h"
+
 namespace Envoy {
 namespace Extensions {
 namespace Http {
 namespace StatefulSession {
 namespace Header {
 
+constexpr absl::string_view OriginUpstreamValuePartFlag = "UV:";
+
 void HeaderBasedSessionStateFactory::SessionStateImpl::onUpdate(
-    const Upstream::HostDescription& host, Envoy::Http::ResponseHeaderMap& headers) {
-  absl::string_view host_address = host.address()->asStringView();
+    absl::string_view host_address, Envoy::Http::ResponseHeaderMap& headers) {
+
+  if (factory_.envelope_mode_) {
+    const auto upstream_value_header = headers.get(factory_.name_);
+    if (upstream_value_header.size() != 1) {
+      ENVOY_LOG(trace, "Header {} not exist or occurs multiple times", factory_.name_);
+      return;
+    }
+
+    const std::string new_header =
+        absl::StrCat(Envoy::Base64::encode(host_address), ";", OriginUpstreamValuePartFlag,
+                     Envoy::Base64::encode(upstream_value_header[0]->value().getStringView()));
+    headers.setReferenceKey(factory_.name_, new_header);
+    return;
+  }
+
+  // Normal OVERRIDE mode.
   if (!upstream_address_.has_value() || host_address != upstream_address_.value()) {
-    const std::string encoded_address =
-        Envoy::Base64::encode(host_address.data(), host_address.length());
-    headers.setCopy(factory_.getHeaderName(), encoded_address);
+    headers.setReferenceKey(factory_.name_, Envoy::Base64::encode(host_address));
   }
 }
 
 HeaderBasedSessionStateFactory::HeaderBasedSessionStateFactory(
     const HeaderBasedSessionStateProto& config)
-    : name_(config.name()) {
+    : name_(config.name()), envelope_mode_(config.mode().has_envelope_mode()) {
   if (config.name().empty()) {
     throw EnvoyException("Header name cannot be empty for header based stateful sessions");
   }
 }
 
+absl::optional<std::string>
+HeaderBasedSessionStateFactory::parseAddress(Envoy::Http::RequestHeaderMap& headers) const {
+  const auto hdr = headers.get(name_);
+  if (hdr.empty()) {
+    return absl::nullopt;
+  }
+  const absl::InlinedVector<absl::string_view, 2> parts =
+      absl::StrSplit(hdr[0]->value().getStringView(), ';', absl::SkipEmpty());
+  if (parts.empty()) {
+    return absl::nullopt;
+  }
+  std::string upstream_host = Base64::decode(parts[0]);
+
+  absl::string_view upstream_value;
+  for (size_t i = 1; i < parts.size(); ++i) {
```

Comment:
Consider iterating from the end backwards, and stop after the first time the if statement is true.

I understand that the current code is better for the general use-case when there will be multiple values to compare (although starting at the end for these will also make more sense), so I'm not strongly against the current form, but it just seems to be better from security perspective (less cpu cycles spent on unused substr operations).

Reply:
I changed it to range-for with a span and break if the `UV:` is found.